app/src/components/generative/ModelMenu.tsx

```diff
@@ -39,19 +40,120 @@ const menuWidthCSS = css`
 export type ModelMenuValue = {
   provider: GenerativeProviderKey;
   modelName: string;
+  /**
+   * The custom provider ID if using a custom provider
+   */
+  customProviderId?: string;
 };
 
+type CustomProviderInfo = {
+  id: string;
+  name: string;
+  sdk: GenerativeModelSDK;
+  modelNames: readonly string[];
+};
+
+/**
+ * Maps GenerativeModelSDK to GenerativeProviderKey for icon display
+ */
+const SDK_TO_PROVIDER_KEY: Record<GenerativeModelSDK, GenerativeProviderKey> = {
+  OPENAI: "OPENAI",
+  AZURE_OPENAI: "AZURE_OPENAI",
+  ANTHROPIC: "ANTHROPIC",
+  AWS_BEDROCK: "AWS",
+  GOOGLE_GENAI: "GOOGLE",
+};
+
+/**
+ * Delimiter used to separate parts of menu keys.
+ * Using a Private Use Area (PUA) character unlikely to appear in provider IDs or model names.
+ */
+const KEY_DELIMITER = "\uE000";
+
+type MenuKeyBuiltIn = {
+  type: "builtin";
+  providerKey: string;
+  modelName: string;
+};
```

Comment:
I would argue the key is the string, this is the encoded model information? It might be more clear if you had

type ModeInfo = BuiltinModelInfo | CustomModelInfo;